### PR TITLE
`[Exiled::Events]` Adding `AddingObserver` and `RemovingObserver` events for Scp173

### PR DESCRIPTION
**Pull request by skyyt15**
the adding observer is triggered when a user just saw scp 173. for example, if ev.IsAllowed is set to false and the user is the only one who can saw scp173, the peanut can keep moving (it will be not added in the Observers list)

an example when the ev.IsAllowed is set to false
image

the RemoveObserver is the same as the AddingObserver event but there is no a IsAllowed parameter (it's weird in this situation) and it's trigger when the player no longer seen scp173.

special thanks to Bolton who help me to find stupid IL errors
special thing to yamato: i have compile this version in release build :trollface:

edit: ho i forgot to tell: i have build all change and tested the events with a different plugin in my personnal server

### CODE REVIEW

**Comment by skyyt15** [comment]
i think it's good now

File: Exiled.Events/Patches/Events/Scp173/Observers.cs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstruction> instructions, ILGenerator generator)
+        {
+            List<CodeInstruction> newInstructions = ListPool<CodeInstruction>.Pool.Get(instructions);
+
+            // AddingObserver patch
+            Label returnLabel = generator.DefineLabel();
+            LocalBuilder ev = generator.DeclareLocal(typeof(AddingObserverEventArgs));
+
+            int index = newInstructions.FindIndex(x => x.Calls(Method(typeof(HashSet<ReferenceHub>), nameof(HashSet<ReferenceHub>.Add)))) + 2;
+            newInstructions[index].WithLabels(returnLabel);
```
Comment (NaoUnderscore) [inline Exiled.Events/Patches/Events/Scp173/Observers.cs:40]:
```suggestion
            newInstructions[index].labels.Add(returnLabel);
```

File: Exiled.Events/EventArgs/Scp173/RemoveObserverEventArgs.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+            Scp173 = player.Role.As<Scp173Role>();
+            Player = player;
+            Target = target;
+        }
+
+        /// <inheritdoc />
+        public Scp173Role Scp173 { get; }
+
+        /// <summary>
+        /// Gets the player who controlling scp 173.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Scp173/RemoveObserverEventArgs.cs:39]:
```suggestion
        /// Gets the player who's controlling SCP-173.
```

File: Exiled.Events/EventArgs/Scp173/RemoveObserverEventArgs.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        /// </summary>
+        public Player Player { get; }
+
+        /// <summary>
+        /// Gets the target who no longer see the scp 173.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Scp173/RemoveObserverEventArgs.cs:44]:
```suggestion
        /// Gets the target who no longer sees SCP-173.
```

File: Exiled.Events/EventArgs/Scp173/AddingObserverEventArgs.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        /// <inheritdoc />
+        public Scp173Role Scp173 { get; }
+
+        /// <summary>
+        /// Gets the target who have just seen SCP-173.
+        /// </summary>
+        public Player Target { get; }
+
+        /// <summary>
+        /// Gets the player who's controlling scp 173.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Scp173/AddingObserverEventArgs.cs:48]:
```suggestion
        /// Gets the player who's controlling SCP-173.
```

File: Exiled.Events/EventArgs/Scp173/AddingObserverEventArgs.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            Player = player;
+            Target = target;
+            IsAllowed = isAllowed;
+        }
+
+        /// <inheritdoc />
+        public Scp173Role Scp173 { get; }
+
+        /// <summary>
+        /// Gets the target who have just seen SCP-173.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Scp173/AddingObserverEventArgs.cs:43]:
```suggestion
        /// Gets the target who has looked at SCP-173.
```

**Review by NaoUnderscore — CHANGES_REQUESTED** [review]
Maintain all `<param name=>`, and such things, on the same line.

**Review by NaoUnderscore — APPROVED** [review]

**Comment by NaoUnderscore** [comment]
Fix build errors

**Pull request merged**